src/shell/shell_misc.cpp

```diff
@@ -389,106 +392,39 @@ void DOS_Shell::InputCommand(char * line) {
-
-	// TODO: Replace this with a regular expression
-	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-	// In the provided 'line' string, replace all instances of %variable_name%s
-	// with equally named environment "variable_name"s'values, as fetched from
-	// the GetEnvStr(variable_name, variable_string_value) function.
-
-	char temp[CMD_MAXLINE]; /* <- NTS: Currently 4096 which is very generous indeed! */
-	char* w = temp;
-	const char* wf = temp + sizeof(temp) - 1;
-	char *r=line;
-	/* initial scan: is there anything to substitute? */
-	/* if not, then return without modifying "line" */
-	while (*r != 0 && *r != '%') r++;
-	if (*r != '%') return;
-	/* if the incoming string is already too long, then that's a problem too! */
-	if (((size_t)(r+1-line)) >= CMD_MAXLINE) {
-		LOG_MSG("DOS_Shell::ProcessCmdLineEnvVarStitution WARNING incoming string to substitute is already too long!\n");
-		goto overflow;
-	}
-	/* copy the string down up to that point */
-	for (const char* c = line; c < r;) {
-		assert(w < wf);
-		*w++ = *c++;
-	}
-	/* begin substitution process */
-	while (*r != 0) {
-		if (*r == '%') {
-			r++;
-			if (*r == '%' || *r == 0) {
-				/* %% or leaving a trailing % at the end (Win95 behavior) becomes a single '%' */
-				if (w >= wf) goto overflow;
-				*w++ = '%';
-				if (*r != 0) r++;
-				else break;
-			}
-			else {
-				const char* name = r; /* store pointer, 'r' is first char of the name following '%' */
-				int spaces = 0,chars = 0;
-				if (isalpha(*r) || *r == ' ') { /* must start with a letter. space is apparently valid too. (Win95) */
-					if (*r == ' ') spaces++;
-					else if (isalpha(*r)) chars++;
-					r++;
-					while (*r != 0 && *r != '%') {
-						if (*r == ' ') spaces++;
-						else chars++;
-						r++;
-					}
-				}
-				if (*r == '%' && ((spaces > 0 && chars == 0) || (spaces == 0 && chars > 0))) {
-					std::string temp2;
-					/* valid name found. substitute */
-					*r++ = 0; /* ASCIIZ snip */
-					if (GetEnvStr(name,temp2)) {
-						size_t equ_pos = temp2.find_first_of('=');
-						if (equ_pos != std::string::npos) {
-							const char *base = temp2.c_str();
-							const char *value = base + equ_pos + 1;
-							const char *fence = base + temp2.length();
-							assert(value >= base && value <= fence);
-							size_t len = (size_t)(fence-value);
-							if ((w+len) > wf) goto overflow;
-							memcpy(w,value,len);
-							w += len;
-						}
-					}
-				}
-				else {
-					/* nope. didn't find a valid name */
-					while (*r == ' ') r++; /* skip spaces */
-					name--; /* step "name" back to cover the first '%' we found */
-					for (const char* c = name; c < r;) {
-						if (w >= wf) goto overflow;
-						*w++ = *c++;
-					}
-				}
-			}
+/* Note: Buffer pointed to by "line" must be at least CMD_MAXLINE+1 bytes long! */
+void DOS_Shell::ProcessCmdLineEnvVarStitution(char* line) {
+	std::string text = line;
+	const static std::regex re("\\%([^%0-9][^%]*)?%");
```

Comment:
Can `constexpr char surrogate_percent = 8;` be moved inside the function and not be global? 
(unless there's a use for it in other shell_misc functions?)

Reply:
Sure. It is not used by any other shell_misc functions at this time, so I moved it inside the function in the code.